### PR TITLE
Adds the ability to walk through a directory tree

### PR DESCRIPTION
**Pull request by johnepiscopo**
This will allow the caller to traverse the directory tree of the remote server

### CODE REVIEW

File: walker.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+type item struct {
+	path  string
+	entry Entry
+	err   error
+}
+
+// Step advances the Walker to the next file or directory,
+// which will then be available through the Path, Stat, and Err methods.
+// It returns false when the walk stops at the end of the tree.
+func (w *Walker) Step() bool {
```
Comment (corny) [inline walker.go]: Better call the function `Next()` like many other iterable objects in go.

Comment (johnepiscopo) [reply walker.go]: Made the change before I saw your edit :-). I think that whilst `Next()` is the right name to use for iterable objects, `Step()` does seem to be the name for directory walking so for consistency I think `Step()` should be used

**Comment by Techassi** [comment]
Any update on the status of this PR?

In a current project I need to walk the directories, but it seems like this PR isn't getting merged. If this is the case I will have to use `goftp` as it provides a `Walk` function.

Cheers

**Comment by Techassi** [comment]
Thanks for the approval!

The test all check out and all suggested changes have been implemented - so will these commits be merged in the near future?

Just asking for an update.

Cheers

**Comment by johnepiscopo** [comment]
Thanks for approving @corny and glad I could contribute! Do you have any idea when these commits would be merged?

**Comment by nicksherron** [comment]
I'd love to use this package for a project I'm working on but can't without this functionality. Any chance we can get this merged?

**Comment by corny** [comment]
@jlaffaye ping

File: walker.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// Next advances the Walker to the next file or directory,
+// which will then be available through the Path, Stat, and Err methods.
+// It returns false when the walk stops at the end of the tree.
+func (w *Walker) Next() bool {
+	if w.descend && w.cur.err == nil && w.cur.entry.Type == EntryTypeFolder {
+		list, err := w.serverConn.List(w.cur.path)
+		if err != nil {
+			w.cur.err = nil
```
Comment (jlaffaye) [inline walker.go]: Why do you ignore the error ?

File: walker.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			w.stack = append(w.stack, w.cur)
+		} else {
+			for i := len(list) - 1; i >= 0; i-- {
+				if !strings.HasSuffix(w.cur.path, "/") {
+					w.cur.path += "/"
+				}
+
+				var path string
+				if list[i].Type == EntryTypeFolder {
+					path = fmt.Sprintf("%s%s", w.cur.path, list[i].Name)
```
Comment (jlaffaye) [inline walker.go]: Its better to use `path.Join` than Sprintf

File: walker.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+type item struct {
+	path  string
+	entry Entry
+	err   error
+}
+
+// Next advances the Walker to the next file or directory,
+// which will then be available through the Path, Stat, and Err methods.
+// It returns false when the walk stops at the end of the tree.
+func (w *Walker) Next() bool {
```
Comment (jlaffaye) [inline walker.go]: Why not return the item ?

File: walker.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	"strings"
+)
+
+//Walker traverses the directory tree of a remote FTP server
+type Walker struct {
+	serverConn *ServerConn
+	root       string
+	cur        item
+	stack      []item
+	descend    bool
```
Comment (jlaffaye) [inline walker.go]: It feels like the boolean should be on the item, as it is linked to the folder

File: walker.go (hunk 2 above)
Comment (johnepiscopo) [reply walker.go]: That's a typo... should be 
`w.cur.err = err`

File: walker.go (hunk 3 above)
Comment (johnepiscopo) [reply walker.go]: Ok

File: walker.go (hunk 4 above)
Comment (johnepiscopo) [reply walker.go]: Was aiming for parity with the Go standard library; their Walker.Step() function returns a bool and I wanted to keep the usage of this implementation as close to that as possible. 

It also means that the caller can only use the Walker object to manage the traversing of the directory tree and the encapsulation of the item within the Walker object maintains a clear separation between the caller and the package.

File: walker.go (hunk 5 above)
Comment (johnepiscopo) [reply walker.go]: Again, I was aiming for parity between this and the Walker implementation from the standard library where the descend is also on the Walker struct. This allows the caller to only have to use the Walker object to traverse the directory tree whilst maintaining a clear separation between the caller and the package.

**Comment by johnepiscopo** [comment]
Thanks, @jlaffaye for taking a look and for maintaining a really useful package

**Pull request merged**